cores/esp8266/core_esp8266_waveform.c

```diff
@@ -149,122 +104,153 @@ void setTimer1Callback(uint32_t (*fn)()) {
-  ets_intr_lock();
-
-  wave->nextTimeHighCycles = MicrosecondsToCycles(timeHighUS) - 70;  // Take out some time for IRQ codepath
-  wave->nextTimeLowCycles = MicrosecondsToCycles(timeLowUS) - 70;  // Take out some time for IRQ codepath
-  wave->timeLeftCycles = MicrosecondsToCycles(runTimeUS);
-  if (!wave->enabled) {
-    wave->state = 0;
+  uint32_t mask = 1<<pin;
+  if (!(waveformEnabled & mask)) {
     // Actually set the pin high or low in the IRQ service to guarantee times
     wave->nextServiceCycle = GetCycleCount() + MicrosecondsToCycles(1);
-    wave->enabled = 1;
+    waveformToEnable |= mask;
     if (!timerRunning) {
       initTimer();
     }
-    ReloadTimer(MicrosecondsToCycles(1)); // Cause an interrupt post-haste
+    timer1_write(MicrosecondsToCycles(10)); // Cause an interrupt post-haste
+    while (waveformToEnable) {
+      delay(0); // Wait for waveform to update
+    }
   }
 
-  // Re-enable interrupts here since we're done with the update
-  ets_intr_unlock();
-
   return true;
 }
 
+// Speed critical bits
+#pragma GCC optimize ("O2")
+// Normally would not want two copies like this, but due to different
+// optimization levels the inline attribute gets lost if we try the
+// other version.
+
+static inline ICACHE_RAM_ATTR uint32_t GetCycleCountIRQ() {
+  uint32_t ccount;
+  __asm__ __volatile__("rsr %0,ccount":"=a"(ccount));
+  return ccount;
+}
+
+static inline ICACHE_RAM_ATTR uint32_t min_u32(uint32_t a, uint32_t b) {
+  if (a < b) {
+    return a;
+  }
+  return b;
+}
+
 // Stops a waveform on a pin
 int ICACHE_RAM_ATTR stopWaveform(uint8_t pin) {
   // Can't possibly need to stop anything if there is no timer active
   if (!timerRunning) {
     return false;
   }
-
-  for (size_t i = 0; i < countof(waveform); i++) {
-    if (!waveform[i].enabled) {
-      continue; // Skip fast to next one, can't need to stop this one since it's not running
-    }
-    if (((pin == 16) && waveform[i].gpio16Mask) || ((pin != 16) && (waveform[i].gpioMask == 1<<pin))) {
-      // Note that there is no interrupt unsafety here.  The IRQ can only ever change .enabled from 1->0
-      // We're also doing that, so even if an IRQ occurred it would still stay as 0.
-      waveform[i].enabled = 0;
-      int cnt = timer1CB ? 1 : 0;
-      for (size_t i = 0; (cnt == 0) && (i < countof(waveform)); i++) {
-        cnt += waveform[i].enabled ? 1 : 0;
-      }
-      if (!cnt) {
-        deinitTimer();
-      }
-      return true;
-    }
+  // If user sends in a pin >16 but <32, this will always point to a 0 bit
+  // If they send >=32, then the shift will result in 0 and it will also return false
+  uint32_t mask = 1<<pin;
+  if (!(waveformEnabled & mask)) {
+    return false; //It's not running, nothing to do here
   }
-  return false;
+  waveformToDisable |= mask;
+  // Ensure tiomely service....
+  if (T1L > MicrosecondsToCycles(10)) {
+    timer1_write(MicrosecondsToCycles(10));
+  }
+  while (waveformToDisable) {
```

Comment:
Does it make sense to wait with timeout?

Reply:
Good eyes, but at this point we know the timer is running and it has, at most, 10us of runtime before the NMI will be called.  So no need to have a timeout here.